agg-ast/schema_derivation/src/schema_derivation_tests/stage.rs

```diff
@@ -1347,6 +1503,84 @@ mod project {
             ..Default::default()
         })
     );
+    test_derive_stage_schema!(
+        project_include_exclude,
+        expected = Ok(Schema::Document(Document {
+            keys: map! {
+                "foo".to_string() => Schema::Array(Box::new(Schema::Document(Document {
+                    keys: map! {
+                        "b".to_string() => Schema::Atomic(Atomic::Integer),
+                    },
+                    ..Default::default()
+                })))
+            },
+            required: set!("foo".to_string()),
+            ..Default::default()
+        })),
+        input = r#"{"$project": {"foo.a": 1, "foo.a": 0}}"#,
+        ref_schema = Schema::Array(Box::new(Schema::Document(Document {
+            keys: map! {
+                "a".to_string() => Schema::Atomic(Atomic::String),
+                "b".to_string() => Schema::Atomic(Atomic::Integer),
+            },
+            ..Default::default()
+        })))
+    );
+    test_derive_stage_schema!(
+        project_exclude_include,
+        expected = Ok(Schema::Document(Document {
+            keys: map! {
+                "foo".to_string() => Schema::Array(Box::new(Schema::Document(Document {
+                    keys: map! {
+                        "a".to_string() => Schema::Atomic(Atomic::String),
+                    },
+                    ..Default::default()
+                })))
+            },
+            required: set!("foo".to_string()),
+            ..Default::default()
+        })),
+        input = r#"{"$project": {"foo.a": 0, "foo.a": 1}}"#,
```

Comment:
Reviewing tests before implementation so I may revisit this comment, but wow this test and the previous one surprise me. The order of the exclude and include matter? Whichever you specify last is the one that takes effect? MongoDB is so weird.

Reply:
it is pretty annoying! we actually just overwrite during deserialization _anyways_, so this should work out of the box. But it revealed the issue with not handling `remove_field` with arrays properly 😄 s/o @terakilobyte for the test suggestion